### PR TITLE
Add Dockerfile and update to Gradle 7

### PR DESCRIPTION
**Pull request by Prof-Bloodstone**
Multi-Stage building to reduce the size
Spring system user

It comes with sane, containerized defaults and allows for overriding configuration through both ENV vars as well as config files:

```sh
docker run --rm \
  --env SERVER_PORT=6969 \
  -v "${PWD}/application.yaml:/app/application.yaml:ro" \
  bibliothek:latest
```

Not sure where would be the best place to put the `dockerdefaults.yaml` file in the source tree.

Open to questions/suggestions :)

Wanted to handle `docker-compose.yaml` while I'm at it, but I noticed there's no Mongo configuration (it always connects to `localhost:27017`), so I left it out for now.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+FROM adoptopenjdk:${JAVA_VERSION}-jre-${JVM_FLAVOR}
+WORKDIR /app
+
+RUN groupadd --system spring \
+    && useradd --system spring --gid spring \
+    && chown -R spring:spring /app
+USER spring:spring
+
+# TODO: Extract to local file?
```
Comment (kashike) [inline Dockerfile:27]: Yes. In proper YAML format.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+RUN printf '%s\n' '{\
+    "app": {\
+        "storagePath": "/data",\
```
Comment (kashike) [inline Dockerfile:30]: `/data/storage`

**Review by kashike — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,39 @@
+ARG JAVA_VERSION=16
+ARG JVM_FLAVOR=hotspot
+
+FROM adoptopenjdk:${JAVA_VERSION}-jdk-${JVM_FLAVOR} AS builder
+WORKDIR /builddir
+
+COPY gradlew build.gradle settings.gradle ./
```
Comment (Proximyst) [inline Dockerfile:7]: Why is this all manual instead of adding the entire directory and instead adding known unnecessary directories to `.dockerignore`?

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+RUN ./gradlew clean buildForDocker --no-daemon
+
+
+ARG JAVA_VERSION
+ARG JVM_FLAVOR
+
+FROM adoptopenjdk:${JAVA_VERSION}-jre-${JVM_FLAVOR}
+WORKDIR /app
+
+RUN groupadd --system spring \
```
Comment (Proximyst) [inline Dockerfile:22]: Commonly one would prefer the app name as group & user (i.e. `bibliothek:bibliothek`), but not a blocker in any sense.

File: Dockerfile (hunk 3 above)
Comment (Prof-Bloodstone) [reply Dockerfile:7]: Did it this way initially to allow for having 2 separate build steps (one for backend and one for CLI), before putting them in the final image.
This also has a few advantages:
Build is not invalidated at all if `dockerdefaults.yaml` is changed (minor benefit)
Gradle wrapper download is not invalidated if only source changes. I'll see if copying `build.gradle` later will work too (I assume it should), as then only changes to the wrapper itself will invalidate it.

I forgot that I should probably support legacy docker builder, which sends whole build context eagerly to docker daemon, so should add `.dockerignore` anyway.

File: docker/dockerdefaults.yaml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+---
+# vim: ft=yaml sw=2
+app:
+  # path to volume in Dockerfile - make sure to keep in sync
+  storagePath: /data/storage
```
Comment (kashike) [inline docker/dockerdefaults.yaml:5]: Double-quote the value.

File: docker/dockerdefaults.yaml [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+  apiBaseUrl: http://localhost/api
```
Comment (kashike) [inline docker/dockerdefaults.yaml:6]: Double-quote the value.

File: docker/dockerdefaults.yaml [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+server:
+  forward-headers-strategy: framework
```
Comment (kashike) [inline docker/dockerdefaults.yaml:8]: Double-quote the value.

File: docker/dockerdefaults.yaml [hunk 8]
```diff
@@ -0,0 +1,9 @@
+---
+# vim: ft=yaml sw=2
```
Comment (kashike) [inline docker/dockerdefaults.yaml:2]: Why is this here?

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+VOLUME /data/storage
+EXPOSE 8080
+
+# We override default config location search path,
+# so that a custom file with defaults can be used
+# Normally would use environment variables,
+# but they take precedence over config file
+# https://docs.spring.io/spring-boot/docs/1.5.6.RELEASE/reference/html/boot-features-external-config.html
+ENV SPRING_CONFIG_LOCATION="optional:classpath:/,optional:classpath:/config/,file:./dockerdefaults.yaml,optional:file:./,optional:file:./config/"
+COPY ./docker/dockerdefaults.yaml ./dockerdefaults.yaml
```
Comment (kashike) [inline Dockerfile:36]: I would rather the file name be something like `default.application.yaml`.

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+ARG JAVA_VERSION
+ARG JVM_FLAVOR
+
+FROM adoptopenjdk:${JAVA_VERSION}-jre-${JVM_FLAVOR}
+WORKDIR /app
+
+RUN groupadd --system bibliothek \
+    && useradd --system bibliothek --gid bibliothek \
+    && chown -R bibliothek:bibliothek /app
+USER spring:spring
```
Comment (kashike) [inline Dockerfile:25]: Tsk.

File: Dockerfile [hunk 11]
```diff
@@ -0,0 +1,39 @@
+ARG JAVA_VERSION=16
+ARG JVM_FLAVOR=hotspot
+
+FROM adoptopenjdk:${JAVA_VERSION}-jdk-${JVM_FLAVOR} AS builder
+WORKDIR /builddir
+
+COPY gradlew ./
+COPY gradle/ gradle/
```
Comment (kashike) [inline Dockerfile:8]: Can't we just `COPY . ./`?

File: Dockerfile [hunk 12]
```diff
@@ -0,0 +1,39 @@
+ARG JAVA_VERSION=16
+ARG JVM_FLAVOR=hotspot
+
+FROM adoptopenjdk:${JAVA_VERSION}-jdk-${JVM_FLAVOR} AS builder
+WORKDIR /builddir
```
Comment (kashike) [inline Dockerfile:5]:
```suggestion
WORKDIR /build
```

**Review by kashike — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 11 above)
Comment (Prof-Bloodstone) [reply Dockerfile:8]: See: https://github.com/PaperMC/bibliothek/pull/43#discussion_r646137550

File: Dockerfile (hunk 9 above)
Comment (Prof-Bloodstone) [reply Dockerfile:36]: Sure. Wanted to avoid doing something like `application-default.yaml` because it seems like that's the file Spring normally looks for given no profile is enabled, but this should be fine.

File: docker/dockerdefaults.yaml (hunk 8 above)
Comment (Prof-Bloodstone) [reply docker/dockerdefaults.yaml:2]: AFAIK most editors parse this header, wanted to mark the width at 2 spaces as in your other file. But I believe it can also be infered, so I'll remove it.

File: Dockerfile (hunk 3 above)
Comment (kashike) [reply Dockerfile]: Would prefer we just add the whole directory. CLI won't be part of this image, so it is not a concern.

Comment (Prof-Bloodstone) [reply Dockerfile]: Was mostly meaning the part where the gradle wrapper is not re-downloaded if it's not required to do so.
I guess I might be optimizing it too much for work flows I know and if Docker is only used for deployment from time to time, it shouldn't matter too much.
Just wanted to avoid pointless minutes wasted waiting for it to download and initialize every time :)

Comment (kashike) [reply Dockerfile]: Yes - please modify to use whole directory as mentioned.

Comment (Prof-Bloodstone) [reply Dockerfile]: Changed

**Review by kashike — APPROVED** [review]

**Review by Proximyst — APPROVED** [review]
